privacyidea/lib/tokens/certificatetoken.py

```diff
@@ -353,6 +359,41 @@ def get_default_settings(cls, g, params):
 
         return ret
 
+    def _update_rollout_state(self):
+        """
+        This is a certificate specific method, that communicates to the CA and checks,
+        if a pending certificate has been enrolled, yet.
+        If the certificate is enrolled, it fetches the certificate from the CA and
+        updates the certificate token.
+        :return:
+        """
+        status = -1
+        if self.rollout_state == ROLLOUTSTATE.PENDING:
+            request_id = self.get_tokeninfo(REQUEST_ID)
+            ca = self.get_tokeninfo("CA")
+            if ca and request_id:
+                request_id = int(request_id)
+                cacon = get_caconnector_object(ca)
+                status = cacon.get_cr_status(request_id)
+                # TODO: Later we need to make the status CA dependent. Different CAs could return
+                # different codes. So each CA Connector needs a mapper for its specific codes.
+                if status in [3, 4]:  # issued or "issued out of band"
+                    log.info("The certificate {0!s} has been issued by the CA.".format(self.token.serial))
+                    certificate = cacon.get_issued_certificate(request_id)
+                    # Update the rollout state
+                    self.token.rollout_state = ROLLOUTSTATE.ENROLLED
+                    self.add_tokeninfo("certificate", certificate)
```

Comment:
maybe we should change the rollout-state **after** the certificate was successfully added.

Reply:
I implemented it this way so I do not need an explicit `self.token.save()` like in the denied state. Since the `self.add_tokeninfo` has a save call.